src/Data/Job.php

```diff
@@ -325,6 +334,84 @@ public function getDeleteAfterDateTime(): \DateTime
         return new \DateTime($deleteAfterDateTimeString);
     }
 
+    public function save(): Db\ModelInterface
+    {
+        if (
+            // if state didn't change, we don't have to override any behavior
+            !array_key_exists(JobInterface::FIELD_NAME_ASSIGNED_STATE, $this->_readChangedPersistentProperties()) ||
+            // because of how State\Service works, often assigned state will show up as changed, when really it hasn't
+            $this->getAssignedState() === $this->getPreviousState()
+        ) {
+            return parent::save();
+        }
+
+        // Db\Models have to use a workaround to get RDBMS bigints to work with doctrine
+        // this code extends that workaround to apply to the job we're about to serialize
+        if ($this->hasId()) {
+            $this->_createOrUpdatePersistentProperty($this->getIdPropertyName(), $this->getId());
+        }
+
+        // do this outside the transaction to marginally reduce the chance of failure
+        $jobStateChange = $this->createJobStateChange();
+
+        $connection = $this->_getDoctrineConnectionDecoratorRepository()->getConnection(DecoratorInterface::ID_JOB);
+        /** @var \PDO $pdo */
+        $pdo = $connection->getWrappedConnection();
+
+        $isKojospaceTransaction = false;
+
+        if ($pdo->inTransaction()) {
+            // if we're already in a transaction (i.e. one started by userspace), we can proceed, knowing that
+            // if userspace rolls back the transaction, it'll roll back the job state change as well
+        } else {
+            // otherwise, start a transaction in kojospace to make sure the JobStateChangelogProcessor doesn't see
+            // the JobStateChange unless the actual job transitions successfully
+            $pdo->beginTransaction();
+            $isKojospaceTransaction = true;
+        }
+
+        try {
+            parent::save();
+            $this->getJobStateChangeRepository()->insertUsingConnection($jobStateChange, $connection);
+
+            if ($isKojospaceTransaction) {
+                $pdo->commit();
+            }
+        } catch (\Throwable $throwable) {
+            if ($isKojospaceTransaction) {
+                $pdo->rollBack();
+            } else {
+                // userspace will need to roll back the transaction itself
+                // if it chooses to commit instead, we might not log this job state change
```

Comment:
From my experience the userspace PDO will throw errors about about and aborted transaction and nothing updated during the transaction will be applied.

```sql
sql> begin
[2019-10-31 11:02:21] completed in 3 ms
sql> select count(*) from kojo_job where type_code = 'auto_area_creation_zip'
[2019-10-31 11:02:25] 1 row retrieved starting from 1 in 31 ms (execution: 12 ms, fetching: 19 ms)
sql> select count(*) from kojo_job where type_code = 'auto_area_creation_zip'
[2019-10-31 11:02:51] 1 row retrieved starting from 1 in 28 ms (execution: 8 ms, fetching: 20 ms)
sql> update kojo_job set type_code = 'non-existing-type-code' where type_code = 'auto_area_creation_zip'
[2019-10-31 11:02:54] [23503] ERROR: insert or update on table "kojo_job" violates foreign key constraint "fk_b0a7b536a01af590"
[2019-10-31 11:02:54] Detail: Key (type_code)=(non-existing-type-code) is not present in table "kojo_job_type".
sql> update kojo_job set type_code = 'auto_area_creation_city' where type_code = 'auto_area_creation_zip'
[2019-10-31 11:03:01] [25P02] ERROR: current transaction is aborted, commands ignored until end of transaction block
[2019-10-31 11:03:01] [23503] ERROR: insert or update on table "kojo_job" violates foreign key constraint "fk_b0a7b536a01af590"
[2019-10-31 11:03:01] Detail: Key (type_code)=(non-existing-type-code) is not present in table "kojo_job_type".
sql> select count(*) from kojo_job where type_code = 'auto_area_creation_zip'
[2019-10-31 11:03:07] [25P02] ERROR: current transaction is aborted, commands ignored until end of transaction block
[2019-10-31 11:03:07] [23503] ERROR: insert or update on table "kojo_job" violates foreign key constraint "fk_b0a7b536a01af590"
[2019-10-31 11:03:07] Detail: Key (type_code)=(non-existing-type-code) is not present in table "kojo_job_type".
sql> commit
```

I think that fact makes this `else` block unnecessary.